doc/man/install-todos.1.in
.TH install\-todos "1"  "Copyright 1996\-2005 FSF" "pilot\-link @VERSION@" 
.SH NAME
install\-todo \- Updates the Palm ToDo list with entries from a local file. 
.SH SECTION
pilot\-link: Userland Conduits
.SH SYNOPSIS
\fIinstall\-todos\fR
[\fB\-p\fR|\fB\-\-port\fR <\fBport\fR>]
[\fB\-\-version\fR] [\fB\-q\fR|\fB\-\-quiet\fR]
[\fB\-?\fR|\fB\-\-help\fR] [\fB\-\-usage\fR]
[\fB\-f\fR|\fB\-\-filename\fR \fBSTRING\fR]
.SH DESCRIPTION
\fIinstall\-todos\fR allows the user to update the Palm
ToDo list with entries from a local file.
.PP
The format of this file is a simple line\-by\-line ToDo task entry.
For each new line in the local file, a new task is created in the
ToDo database on the Palm.
.SH OPTIONS
.SS "INSTALL\-TODOS OPTIONS"
\fB\-f\fR,
\fB\-\-filename\fR \fBfilename\fR
.RS 
A local filename containing the ToDo entry text.
.RE
.SS "CONDUIT OPTIONS"
\fB\-p\fR, \fB\-\-port\fR
<\fBport\fR>
.RS 
Use device file <\fBport\fR> to communicate
with the Palm handheld. If this is not specified,
\fIinstall\-todos\fR will look for the
\fB$PILOTPORT\fR environment variable. If neither
are
found or supplied, \fIinstall\-todos\fR will
print the usage information.
.RE
.PP
\fB\-q\fR, \fB\-\-quiet\fR
.RS 
Suppress 'Hit HotSync button' message
.RE
.PP
\fB\-v\fR, \fB\-\-version\fR
.RS 
Display version of \fIinstall\-todo\fR
and exit without connecting.
.RE
.SS "HELP OPTIONS"
\fB\-h\fR, \fB\-\-help\fR
.RS 
Display the help synopsis for
\fIinstall\-todos\fR and exit without connecting.
.RE
.PP
\fB\-\-usage\fR 
.RS 
Display a brief usage message and exit without connecting.
.RE
.SH USAGE
The program will connect to a target Palm handheld, and creates the
ToDo entry to the Palm ToDo application, using the specified file.
.SH EXAMPLE
\fB\-p\fR
\fB/dev/pilot\fR
\fB\-f\fR
\fB$HOME/MyTodoList.txt\fR
.SH AUTHOR
\fIinstall\-todo\fR was originally written by
Martin Fick, based on code in \fIinstall\-todos\fR by
Robert A. Kaplan.
.PP
This manual page was written by Neil Williams
linux@codehelp.co.uk
.SH "REPORTING BUGS"
Report bugs at:
\fIhttp://bugs.pilot\-link.org\fR
.SH COPYRIGHT
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the 
License, or (at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.
.PP
You should have received a copy of the GNU General Public
License along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111\-1307 
USA
.SH "SEE ALSO"
\fIpilot\-link\fR(7).


src/install-todos.c
/*
 * install-todolist.c:  Palm ToDo list installer
 *
 * Copyright 1996, Robert A. Kaplan
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 */

#include "pi-userland.h"
#include <stdio.h>
#include <stdlib.h>

#include "pi-source.h"
#include "pi-dlp.h"
#include "pi-todo.h"
#include "pi-header.h"
#include "pi-buffer.h"

void install_ToDos(int sd, int db, char *filename)
{
	int 	cLen		= 0,
		i		= 0,
		filelen;

        char 	*file_text 	= NULL,
		*cPtr 		= file_text,
		*begPtr 	= cPtr,
		note_text[] 	= "";

        pi_buffer_t *ToDo_buf;

        struct 	ToDo todo;
        FILE 	*f;

	f = fopen(filename, "r");
	if (f == NULL) {
		perror("fopen");
		exit(EXIT_FAILURE);
	}

	fseek(f, 0, SEEK_END);
	filelen = ftell(f);
	fseek(f, 0, SEEK_SET);

	file_text = (char *) malloc(filelen + 1);
	if (file_text == NULL) {
		perror("malloc()");
		exit(EXIT_FAILURE);
	}

	fread(file_text, filelen, 1, f);

        cPtr = file_text;
        begPtr = cPtr;
        cLen = 0;
        i = 0;
	while (i < filelen) {
		i++;
		/* printf("c:%c.\n",*cPtr); */
		if (*cPtr == '\n') {
			todo.description = begPtr;
			/* replace CR with terminator */
			*cPtr = '\0';

			todo.priority 	= 4;
			todo.complete 	= 0;
			todo.indefinite = 1;
			/* now = time(0);
			   todo.due = *localtime(&now); */
			todo.note = note_text;
			ToDo_buf = pi_buffer_new(0);
			pack_ToDo(&todo, ToDo_buf, todo_v1);
			printf("Description: %s\n", todo.description);

			/* printf("todobuf: %s\n",ToDo_buf->data);       */
			dlp_WriteRecord(sd, db, 0, 0, 0, ToDo_buf->data,
					ToDo_buf->used, 0);
			pi_buffer_free(ToDo_buf);
			cPtr++;
			begPtr = cPtr;
			cLen = 0;
		} else {
			cLen++;
			cPtr++;
		}
	}
	return;
}

int main(int argc, const char *argv[])
{
	int 	c,		/* switch */
		db,
		sd 		= -1;

        const char
                *progname 	= argv[0];

        char 	*filename 	= NULL;

	poptContext po;

	struct poptOption options[] = {
		USERLAND_RESERVED_OPTIONS
	        {"filename",	'f', POPT_ARG_STRING, &filename, 0, "A local file with formatted ToDo entries"},
		POPT_TABLEEND
	};

	po = poptGetContext("install-todos", argc, argv, options, 0);
	poptSetOtherOptionHelp(po,
		"\n\n"
		"   Updates the Palm ToDo list with entries from a local file\n\n"
		"   Example:\n"
		"      -p /dev/pilot -f $HOME/MyTodoList.txt\n\n"
		"   The format of this file is a simple line-by-line ToDo task entry.\n"
		"   For each new line in the local file, a new task is created in the\n"
		"   ToDo database on the Palm.\n\n");

	if (argc<2) {
		poptPrintUsage(po,stderr,0);
		return 1;
	}

	while ((c = poptGetNextOpt(po)) >= 0) {
		fprintf(stderr,"   ERROR: Unhandled option %d.\n",c);
		return 1;
	}

	if (c < -1) {
		plu_badoption(po,c);
	}

	if (filename == NULL) {
		fprintf(stderr,"\n   ERROR: You must specify a filename to read ToDo entries from.\n"
			"   Please see %s --help for more information.\n\n", progname);
		exit(EXIT_FAILURE);
	}


		sd = plu_connect();
		if (sd < 0)
			return 1;

		/* Did we get a valid socket descriptor back? */
		if (dlp_OpenConduit(sd) < 0) {
			pi_close(sd);
			return 1;
		}

		/* Tell user (via Palm) that we are starting things up */
		dlp_OpenConduit(sd);

		/* Open the ToDo Pad's database, store access handle in db */
		if (dlp_OpenDB(sd, 0, 0x80 | 0x40, "ToDoDB", &db) < 0) {
			fprintf(stderr,"   ERROR: Unable to open ToDoDB on Palm.\n");
			dlp_AddSyncLogEntry(sd, "Unable to open ToDoDB.\n");
			pi_close(sd);
			return 1;
		}

		/* Actually do the install here, passed a filename */
		install_ToDos(sd, db, filename);

		/* Close the database */
		dlp_CloseDB(sd, db);

		dlp_AddSyncLogEntry(sd, "Wrote ToDo list entries to Palm.\nThank you for using pilot-link.\n");
		dlp_EndOfSync(sd, 0);
		pi_close(sd);
	return 0;
}
